src/test/java/uk/gov/hmcts/reform/notificationservice/service/NotificationMessageProcessorTest.java

```diff
@@ -0,0 +1,263 @@
+        verify(messageReceiver).complete(lock);
+
+    }
+
+
+    @Test
+    public void should_dead_letter_the_message_when_unrecoverable_failure() throws Exception {
+        // given
+        IMessage message = mock(IMessage.class);
+        MessageBody messageBody = mock(MessageBody.class);
+        given(message.getMessageBody()).willReturn(messageBody);
+
+        UUID lock = UUID.randomUUID();
+        given(message.getLockToken()).willReturn(lock);
+
+        given(messageReceiver.receive()).willReturn(message);
+
+        given(notificationMessageParser.parse(messageBody)).willThrow(new InvalidMessageException("JsonException"));
+
+        // when
+        notificationMessageProcessor.processNextMessage();
+
+        // then
+        verify(messageReceiver).receive();
+
+        verify(messageReceiver).deadLetter(
+            eq(lock),
+            eq("Notification Message processing error"),
+            eq("UNRECOVERABLE_FAILURE"));
+        verifyNoMoreInteractions(messageReceiver);
+    }
+
+
+    @Test
+    public void should_not_dead_letter_the_message_when_recoverable_failure() throws Exception {
+        IMessage message = mock(IMessage.class);
+        MessageBody messageBody = mock(MessageBody.class);
+        given(message.getMessageBody()).willReturn(messageBody);
+
+        given(messageReceiver.receive()).willReturn(message);
+
+        NotificationMsg notificationMsg = mock(NotificationMsg.class);
+        given(notificationMessageParser.parse(messageBody)).willReturn(notificationMsg);
+
+        Exception processingFailureCause = new RuntimeException(
+            "exception of type treated as recoverable"
+        );
+
+        // given an error occurs during message processing
+        willThrow(processingFailureCause)
+            .given(notificationMessageHandler).handleNotificationMessage(notificationMsg);
+        // when
+        notificationMessageProcessor.processNextMessage();
+
+        // then the message is not finalised (completed/dead-lettered)
+        verify(messageReceiver).receive();
+        verifyNoMoreInteractions(messageReceiver);
+
+    }
+
+
+    @Test
+    public void should_dead_letter_the_message_when_recoverable_failure_but_delivery_maxed() throws Exception {
+        // given
+        IMessage message = mock(IMessage.class);
+        MessageBody messageBody = mock(MessageBody.class);
+        given(message.getMessageBody()).willReturn(messageBody);
+        UUID lock = UUID.randomUUID();
+        given(message.getLockToken()).willReturn(lock);
+        given(messageReceiver.receive()).willReturn(message);
+
+        NotificationMsg notificationMsg = mock(NotificationMsg.class);
+        given(notificationMessageParser.parse(messageBody)).willReturn(notificationMsg);
+
+        notificationMessageProcessor = new NotificationMessageProcessor(
+            messageReceiver,
+            notificationMessageHandler,
+            notificationMessageParser,
+            1
```

Comment:
Just a suggestion: as 1 is quite a special value, it would be ideal to set it to something higher and mock the message to return delivery count matching that number. It would also be more explicit this way. Up to you.

Reply:
chnaged, using the main one and mock to delevery count